client/xr/htc_body_tracker.cpp

```diff
@@ -0,0 +1,155 @@
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include "htc_body_tracker.h"
+#include "xr/details/enumerate.h"
+#include "xr/htc_exts.h"
+#include "xr/xr.h"
+#include <openxr/openxr.h>
+
+// Note: This utilizes extensions not present in the OpenXR spec!
+// For more info see:
+// https://hub.vive.com/apidoc/api/VIVE.OpenXR.Tracker.ViveXRTracker.html
+// https://hub.vive.com/apidoc/api/VIVE.OpenXR.VivePathEnumerationHelper.xrEnumeratePathsForInteractionProfileHTCDelegate.html
+
+xr::vive_xr_tracker::vive_xr_tracker(xr::space && space) :
+        space(std::move(space)) {}
+void xr::vive_xr_tracker::set_active(bool active)
+{
+	is_active = active;
+}
+bool xr::vive_xr_tracker::get_active() const
+{
+	return is_active;
+}
+xr::space * xr::vive_xr_tracker::get_space()
+{
+	return &space;
+}
+
+static wivrn::from_headset::body_tracking::pose locate_space(XrSpace space, XrSpace reference, XrTime time)
+{
+	XrSpaceVelocity velocity{
+	        .type = XR_TYPE_SPACE_VELOCITY,
+	};
+
+	XrSpaceLocation location{
+	        .type = XR_TYPE_SPACE_LOCATION,
+	        .next = &velocity,
+	};
+
+	xrLocateSpace(space, reference, time, &location);
+
+	wivrn::from_headset::body_tracking::pose res{
+	        .pose = location.pose,
+	        .flags = 0,
+	};
+
+	if (location.locationFlags & XR_SPACE_LOCATION_ORIENTATION_VALID_BIT)
+		res.flags |= wivrn::from_headset::body_tracking::orientation_valid;
+
+	if (location.locationFlags & XR_SPACE_LOCATION_POSITION_VALID_BIT)
+		res.flags |= wivrn::from_headset::body_tracking::position_valid;
+
+	if (location.locationFlags & XR_SPACE_LOCATION_ORIENTATION_TRACKED_BIT)
+		res.flags |= wivrn::from_headset::body_tracking::orientation_tracked;
+
+	if (location.locationFlags & XR_SPACE_LOCATION_POSITION_TRACKED_BIT)
+		res.flags |= wivrn::from_headset::body_tracking::position_tracked;
+
+	return res;
+}
+
+std::vector<std::string> xr::htc_body_tracker::get_roles()
+{
+	auto paths = get_paths();
+	std::vector<std::string> roles(paths.size());
+
+	for (auto & path: paths)
+	{
+		XrInputSourceLocalizedNameGetInfo get_info{
+		        .type = XR_TYPE_INPUT_SOURCE_LOCALIZED_NAME_GET_INFO,
+		        .next = nullptr,
+		        .sourcePath = path,
+		        .whichComponents = XR_INPUT_SOURCE_LOCALIZED_NAME_USER_PATH_BIT,
+		};
+
+		roles.emplace_back(xr::details::enumerate<char>(xrGetInputSourceLocalizedName, s, &get_info));
```

Comment:
This creates a vector of `path.size()` elements, initialized to the default value (empty string)
```cpp
std::vector<std::string> roles(paths.size());
```
You should instead do
```cpp
std::vector<std::string> roles:
roles.reserve(paths.size());
```
Or do a `roles[i] = ...` instead of `emplace_back`